pubsub/gossipsub/gossipsub-v1.3.md

```diff
@@ -0,0 +1,137 @@
+# gossipsub v1.3: Extensions Control Message
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2025-06-23  |
+
+Authors: [@marcopolo]
+
+Interest Group: TODO
+
+[@marcopolo]: https://github.com/marcopolo
+
+See the [lifecycle document][lifecycle-spec] for context about the maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+## Overview
+
+This version specifies a way to for gossipsub peers to describe their
+characteristics to each other without requiring a new protocol ID per extension.
+
+This spec MUST be updated upon introducing a new extension, either canonical or
+experimental, to the network.
+
+## Motivation
+
+This version makes Gossipsub easier to extend by allowing applications to
+selectively make use of the extensions it would benefit from. It removes the
+need to make Gossipsub extensions follow a strict ordering. Finally, it allows
+extensions to iterate independently from Gossipsub's versioning.
+
+## The Extensions Control Message
+
+If a peer supports any extension, the Extensions control message MUST be
```

Comment:
We should future-proof this in case we move away from the single-stream (per direction) design. For context: we speculate that more granular streams on top of QUIC can alleviate transport HoL blocking and can enable more fluid cancellation. We've discussed:

- one stream per subscription (+ control stream)
- one stream per message (+ control stream)

Reply:
I don't think we need any text change here. We'll always have an initial stream and the extensions MUST be in the first message on that stream.

The best next optimization for multiple streams would be to open a stream per message. I don't think multiple control streams are worth it because it is simpler if control messages are serialized and control messages are small.

Aside, it would be great to use bidirectional streams as bidirectional. There is implementation complexity in correctly treating two separate streams as a single bidirectional stream (the current go implementation has bugs here).